## Objective:
## • Employ SVM from scikit learn for binary classification.
## • Impact of preprocessing data and hyper paramter search using grid search.

In [32]:
## Importing Libraries

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
warnings.filterwarnings('ignore')

### 1. Load the data from “college.csv” that has attributes collected about private and public colleges for a particular year. We will try to predict the private/public status of the college from other attributes.

In [2]:
## Reading data from file into dataframe

df=pd.read_csv(r"./datasets/college.csv")

In [7]:
df.head(10)

Private  Apps  Accept  Enroll  Top10perc  Top25perc  F.Undergrad  \
0     Yes  1660    1232     721         23         52         2885   
1     Yes  2186    1924     512         16         29         2683   
2     Yes  1428    1097     336         22         50         1036   
3     Yes   417     349     137         60         89          510   
4     Yes   193     146      55         16         44          249   
5     Yes   587     479     158         38         62          678   
6     Yes   353     340     103         17         45          416   
7     Yes  1899    1720     489         37         68         1594   
8     Yes  1038     839     227         30         63          973   
9     Yes   582     498     172         21         44          799   

   P.Undergrad  Outstate  Room.Board  Books  Personal  PhD  Terminal  \
0          537      7440        3300    450      2200   70        78   
1         1227     12280        6450    750      1500   29        30   
2           99     11250        3750    400      1165   53        66   
3           63     12960        5450    450       875   92        97   
4          869      7560        4120    800      1500   76        72   
5           41     13500        3335    500       675   67        73   
6          230     13290        5720    500      1500   90        93   
7           32     13868        4826    450       850   89       100   
8          306     15595        4400    300       500   79        84   
9           78     10468        3380    660      1800   40        41   

   S.F.Ratio  perc.alumni  Expend  Grad.Rate  
0       18.1           12    7041         60  
1       12.2           16   10527         56  
2       12.9           30    8735         54  
3        7.7           37   19016         59  
4       11.9            2   10922         15  
5        9.4           11    9727         55  
6       11.5           26    8861         63  
7       13.7           37   11487         73  
8       11.3           23   11644         80  
9       11.5           15    8991         52

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 777 entries, 0 to 776
Data columns (total 18 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Private      777 non-null    object 
 1   Apps         777 non-null    int64  
 2   Accept       777 non-null    int64  
 3   Enroll       777 non-null    int64  
 4   Top10perc    777 non-null    int64  
 5   Top25perc    777 non-null    int64  
 6   F.Undergrad  777 non-null    int64  
 7   P.Undergrad  777 non-null    int64  
 8   Outstate     777 non-null    int64  
 9   Room.Board   777 non-null    int64  
 10  Books        777 non-null    int64  
 11  Personal     777 non-null    int64  
 12  PhD          777 non-null    int64  
 13  Terminal     777 non-null    int64  
 14  S.F.Ratio    777 non-null    float64
 15  perc.alumni  777 non-null    int64  
 16  Expend       777 non-null    int64  
 17  Grad.Rate    777 non-null    int64  
dtypes: float64(1), int64(16), object(1)
memory usage: 

In [5]:
df.describe()

Apps        Accept       Enroll   Top10perc   Top25perc  \
count    777.000000    777.000000   777.000000  777.000000  777.000000   
mean    3001.638353   2018.804376   779.972973   27.558559   55.796654   
std     3870.201484   2451.113971   929.176190   17.640364   19.804778   
min       81.000000     72.000000    35.000000    1.000000    9.000000   
25%      776.000000    604.000000   242.000000   15.000000   41.000000   
50%     1558.000000   1110.000000   434.000000   23.000000   54.000000   
75%     3624.000000   2424.000000   902.000000   35.000000   69.000000   
max    48094.000000  26330.000000  6392.000000   96.000000  100.000000   

        F.Undergrad   P.Undergrad      Outstate   Room.Board        Books  \
count    777.000000    777.000000    777.000000   777.000000   777.000000   
mean    3699.907336    855.298584  10440.669241  4357.526384   549.380952   
std     4850.420531   1522.431887   4023.016484  1096.696416   165.105360   
min      139.000000      1.000000   2340.000000  1780.000000    96.000000   
25%      992.000000     95.000000   7320.000000  3597.000000   470.000000   
50%     1707.000000    353.000000   9990.000000  4200.000000   500.000000   
75%     4005.000000    967.000000  12925.000000  5050.000000   600.000000   
max    31643.000000  21836.000000  21700.000000  8124.000000  2340.000000   

          Personal         PhD    Terminal   S.F.Ratio  perc.alumni  \
count   777.000000  777.000000  777.000000  777.000000   777.000000   
mean   1340.642214   72.660232   79.702703   14.089704    22.743887   
std     677.071454   16.328155   14.722359    3.958349    12.391801   
min     250.000000    8.000000   24.000000    2.500000     0.000000   
25%     850.000000   62.000000   71.000000   11.500000    13.000000   
50%    1200.000000   75.000000   82.000000   13.600000    21.000000   
75%    1700.000000   85.000000   92.000000   16.500000    31.000000   
max    6800.000000  103.000000  100.000000   39.800000    64.000000   

             Expend  Grad.Rate  
count    777.000000  777.00000  
mean    9660.171171   65.46332  
std     5221.768440   17.17771  
min     3186.000000   10.00000  
25%     6751.000000   53.00000  
50%     8377.000000   65.00000  
75%    10830.000000   78.00000  
max    56233.000000  118.00000

In [6]:
df.isnull().sum()

Private        0
Apps           0
Accept         0
Enroll         0
Top10perc      0
Top25perc      0
F.Undergrad    0
P.Undergrad    0
Outstate       0
Room.Board     0
Books          0
Personal       0
PhD            0
Terminal       0
S.F.Ratio      0
perc.alumni    0
Expend         0
Grad.Rate      0
dtype: int64

In [8]:
df["Private"].value_counts()

Yes    565
No     212
Name: Private, dtype: int64

### 2. Use LabelEncoder to encode the target variable in to numerical form and split the data such that 20% of the data is set aside for testing.

In [9]:
## apply get dummies function for encoding
df_result = pd.get_dummies(df,drop_first=True)

In [10]:
df_result.head(10)

Apps  Accept  Enroll  Top10perc  Top25perc  F.Undergrad  P.Undergrad  \
0  1660    1232     721         23         52         2885          537   
1  2186    1924     512         16         29         2683         1227   
2  1428    1097     336         22         50         1036           99   
3   417     349     137         60         89          510           63   
4   193     146      55         16         44          249          869   
5   587     479     158         38         62          678           41   
6   353     340     103         17         45          416          230   
7  1899    1720     489         37         68         1594           32   
8  1038     839     227         30         63          973          306   
9   582     498     172         21         44          799           78   

   Outstate  Room.Board  Books  Personal  PhD  Terminal  S.F.Ratio  \
0      7440        3300    450      2200   70        78       18.1   
1     12280        6450    750      1500   29        30       12.2   
2     11250        3750    400      1165   53        66       12.9   
3     12960        5450    450       875   92        97        7.7   
4      7560        4120    800      1500   76        72       11.9   
5     13500        3335    500       675   67        73        9.4   
6     13290        5720    500      1500   90        93       11.5   
7     13868        4826    450       850   89       100       13.7   
8     15595        4400    300       500   79        84       11.3   
9     10468        3380    660      1800   40        41       11.5   

   perc.alumni  Expend  Grad.Rate  Private_Yes  
0           12    7041         60            1  
1           16   10527         56            1  
2           30    8735         54            1  
3           37   19016         59            1  
4            2   10922         15            1  
5           11    9727         55            1  
6           26    8861         63            1  
7           37   11487         73            1  
8           23   11644         80            1  
9           15    8991         52            1

In [12]:
df_result.rename(columns={"Private_Yes" : 'Private'},inplace=True)

In [13]:
df_result.columns

Index(['Apps', 'Accept', 'Enroll', 'Top10perc', 'Top25perc', 'F.Undergrad',
       'P.Undergrad', 'Outstate', 'Room.Board', 'Books', 'Personal', 'PhD',
       'Terminal', 'S.F.Ratio', 'perc.alumni', 'Expend', 'Grad.Rate',
       'Private'],
      dtype='object')

In [15]:
y_col = ["Private"]
x_col = ['Apps', 'Accept', 'Enroll', 'Top10perc', 'Top25perc', 'F.Undergrad',
       'P.Undergrad', 'Outstate', 'Room.Board', 'Books', 'Personal', 'PhD',
       'Terminal', 'S.F.Ratio', 'perc.alumni', 'Expend', 'Grad.Rate']

In [17]:
X = df_result[x_col]
Y = df_result[y_col]

In [18]:
Y = np.squeeze(Y)

In [19]:
## importing test train split package
from sklearn.model_selection import train_test_split

In [20]:
X_train, X_test, y_train, y_test = train_test_split(X, Y, test_size=0.20,random_state=48, stratify = Y)

In [21]:
print(X_train.shape, X_test.shape, y_train.shape, y_test.shape)

(621, 17) (156, 17) (621,) (156,)


### 3. Fit a linear svm from scikit learn and observe the accuracy.
[Hint: Use Linear SVC]

In [23]:
## Importing Logistic regression Library
from sklearn.svm import LinearSVC

In [30]:
lsvc = LinearSVC(random_state = 48, max_iter = 3000, penalty='l2',C=1)

In [33]:
lsvc.fit(X_train, y_train)

LinearSVC(C=1, max_iter=3000, random_state=48)

In [34]:
y_pred = lsvc.predict(X_test)

In [35]:
from sklearn.metrics import confusion_matrix,classification_report,accuracy_score

In [39]:
print(f"Accuracy score of Model on Training Data = {round(lsvc.score(X_train, y_train),2)} \n")

Accuracy score of Model on Training Data = 0.93 



In [40]:
print(f"Accuracy score of Model on Test Data = {round(accuracy_score(y_test,y_pred),2)} \n\n")

Accuracy score of Model on Test Data = 0.88 




In [41]:
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.93      0.60      0.73        43
           1       0.87      0.98      0.92       113

    accuracy                           0.88       156
   macro avg       0.90      0.79      0.83       156
weighted avg       0.88      0.88      0.87       156



### 4. Preprocess the data using StandardScalar and fit the same model again and observe the change in accuracy.
[Hint: Refer to scikitlearn’s preprocessing methods]

In [42]:
## Apply Standardization 
from sklearn.preprocessing import StandardScaler

In [43]:
scaler = StandardScaler()
# Fit on training set only.
scaler.fit(X_train)

StandardScaler()

In [44]:
# Apply transform to both the training set and the test set.
X_train = scaler.transform(X_train)
X_test = scaler.transform(X_test)

#### Fitting the model again

In [51]:
lsvc_2 = LinearSVC(random_state = 48, max_iter = 3000, penalty='l2',C=1)

In [52]:
lsvc_2.fit(X_train, y_train)

LinearSVC(C=1, max_iter=3000, random_state=48)

In [53]:
y_pred = lsvc_2.predict(X_test)

In [54]:
print(f"Accuracy score of Model on Training Data = {round(lsvc_2.score(X_train, y_train),2)} \n")

Accuracy score of Model on Training Data = 0.95 



In [55]:
print(f"Accuracy score of Model on Test Data = {round(accuracy_score(y_test,y_pred),2)} \n\n")

Accuracy score of Model on Test Data = 0.93 




In [56]:
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.92      0.81      0.86        43
           1       0.93      0.97      0.95       113

    accuracy                           0.93       156
   macro avg       0.93      0.89      0.91       156
weighted avg       0.93      0.93      0.93       156



### Fitting the model again after Standardization did increased the accuracy of model.

### 5. Use scikit learn’s gridsearch to select the best hyperparameter for a non-linear SVM, identify the model with best score and its parameters.
[Hint: Refer to model_selection module of Scikit learn]

In [62]:
## importing the non-linear model
from sklearn.svm import SVC

In [63]:
from sklearn.model_selection import GridSearchCV

In [80]:
param_grid = {
    'C':[1,1.5,2],
    'kernel':['linear','poly', 'rbf','sigmoid'],
    'gamma':['scale','auto'],
    'max_iter':[100,500,1000,5000],
    'decision_function_shape': ['ovo','ovr']
}

In [81]:
svc = SVC(random_state = 48)

In [82]:
grid = GridSearchCV(estimator=svc, param_grid=param_grid, cv = 5)

In [83]:
grid_results = grid.fit(X_train, y_train)

In [84]:
grid_results.cv_results_

{'mean_fit_time': array([0.00297928, 0.00362897, 0.00359783, 0.00374045, 0.00244303,
        0.00290022, 0.00282941, 0.0027164 , 0.00259943, 0.00296464,
        0.00328746, 0.00307617, 0.00275683, 0.00287991, 0.00300698,
        0.00294509, 0.00163412, 0.00270591, 0.00268035, 0.0028564 ,
        0.0022584 , 0.00272317, 0.00287857, 0.00277367, 0.00290675,
        0.00327077, 0.00323133, 0.00289497, 0.00260496, 0.00302334,
        0.00271482, 0.00258255, 0.00144   , 0.00237899, 0.00248303,
        0.0025702 , 0.00204139, 0.00258021, 0.00256257, 0.00261488,
        0.00262742, 0.00292001, 0.00291314, 0.00287228, 0.00264006,
        0.0026782 , 0.00284948, 0.00336957, 0.00162044, 0.00276022,
        0.002808  , 0.00297065, 0.00233583, 0.00282493, 0.00284142,
        0.00284991, 0.00295   , 0.00315943, 0.00300899, 0.00287333,
        0.00250721, 0.00252504, 0.00257378, 0.00251541, 0.00146456,
        0.00252886, 0.00277953, 0.00319405, 0.00210643, 0.00271897,
        0.00271678, 0.00321851,

In [85]:
grid_results.best_params_

{'C': 1.5,
 'decision_function_shape': 'ovo',
 'gamma': 'scale',
 'kernel': 'linear',
 'max_iter': 500}

### Bulilding Model again with best parameters

In [98]:
svc_final = SVC( C= 1.5,
                 decision_function_shape = 'ovo',
                 gamma ='scale',
                 kernel = 'linear',
                 max_iter = 500,
                 random_state = 48)

In [99]:
svc_final.fit(X_train, y_train)

SVC(C=1.5, decision_function_shape='ovo', kernel='linear', max_iter=500,
    random_state=48)

In [100]:
y_pred = svc_final.predict(X_test)

In [101]:
print(f"Accuracy score of Model on Training Data = {round(svc_final.score(X_train, y_train),2)} \n")

Accuracy score of Model on Training Data = 0.95 



In [102]:
print(f"Accuracy score of Model on Test Data = {round(accuracy_score(y_test,y_pred),2)} \n\n")

Accuracy score of Model on Test Data = 0.93 




In [103]:
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.92      0.81      0.86        43
           1       0.93      0.97      0.95       113

    accuracy                           0.93       156
   macro avg       0.93      0.89      0.91       156
weighted avg       0.93      0.93      0.93       156



## Non-Linear SVC model has same accuracy as LinearSVC
## Both of them have Accuracy of 0.93